lib/Conversion/SeqToSV/FirRegLowering.h

```diff
@@ -16,20 +16,125 @@
 #include "circt/Support/LLVM.h"
 #include "circt/Support/Namespace.h"
 #include "circt/Support/SymCache.h"
+#include <bits/chrono.h>
+#include <chrono>
 
 namespace circt {
+struct ValueSCC {
+  struct SccInfo {
+    size_t order;
+    size_t lowlink;
+    size_t componentId;
+  };
+  llvm::DenseMap<Value, SccInfo> valueInfoMap;
+  unsigned componentIdGen = 0;
+  unsigned index = 0;
+  llvm::SmallVector<Value> stack;
+  llvm::DenseSet<Value> onStack;
+  llvm::function_ref<bool(Operation *)> filter;
+  ValueSCC(hw::HWModuleOp moduleOp,
+           llvm::function_ref<bool(Operation *)> f = nullptr) {
+    if (f)
+      filter = f;
+    auto time_begin = std ::chrono::high_resolution_clock::now();
+    llvm::errs() << "\n begin scc:";
+    for (Operation *rootOp : moduleOp.getOps<seq::FirRegOp>())
+      tarjanSCCiterative(rootOp);
+    auto time_end = std ::chrono::high_resolution_clock::now();
+    llvm::errs() << "\n time to complete scc:"
+                 << std::chrono::duration_cast<std::chrono::seconds>(time_end -
+                                                                     time_begin)
+                        .count();
+  };
+
+  void tarjanSCCiterative(Operation *rootOp) {
+    // Stack to simulate the recursive call stack
+    SmallVector<std::pair<Value, bool>, 64> dfsStack;
+    for (auto val : rootOp->getResults())
+      dfsStack.push_back({val, true});
+
+    llvm::DenseMap<Value, Value> parentOf;
+    while (!dfsStack.empty()) {
+      Value visitVal = dfsStack.back().first;
+      auto &firstVisit = dfsStack.back().second;
+
+      if (firstVisit) {
+        valueInfoMap[visitVal] = {index, index, 0};
+        ++index;
+        stack.push_back(visitVal);
+        onStack.insert(visitVal);
+      }
+
+      bool continueTraversal = false;
+      size_t minLowLink = valueInfoMap[visitVal].lowlink;
+      for (auto *user : visitVal.getUsers()) {
+        // If ops need to be filtered ignore them, cannot ignore rootOp,
+        // otherwise SCC cannot be computed.
+        if (user != rootOp && filter && filter(user))
+          continue;
+        for (auto childVal : user->getResults()) {
+          // set the firstvisit flag of visitVal to false.
+          firstVisit = false;
+          if (!valueInfoMap.contains(childVal)) {
+            // If child not yet visited.
+            dfsStack.push_back(std::make_pair(childVal, true));
+            continueTraversal = true;
+            parentOf[childVal] = visitVal;
+            // Simulate dfs traversal, defer visitVal traversal and start
+            // childVal traversal.
+            break;
+          }
```

Comment:
I kind feel this implementation is not appropriate. This visits all users every time so I think we need to continue iterations from the point where the last visit ends.

Reply:
We could change `dfsStack` to `SmallVector<std::tuple<Value, bool, user_iter>, 64>` but it might be simpler to just rely on scc-iter.